crypto/bio/bss_dgram.c

```diff
@@ -790,6 +978,635 @@ static int dgram_puts(BIO *bp, const char *str)
+                      control[i], &BIO_MSG_N(msg, i));
+
+        /* If local address was requested, it must have been enabled */
+        if (BIO_MSG_N(msg, i).local != NULL && !have_local_enabled)
+            return -1;
+    }
+
+    /* Do the batch */
+    ret = recvmmsg(b->num, mh, num_msg, sysflags, NULL);
+    if (ret > 0) {
+        for (i = 0; i < (size_t)ret; ++i) {
+            BIO_MSG_N(msg, i).data_len = mh[i].msg_len;
+            BIO_MSG_N(msg, i).flags    = 0;
+            /*
+             * *(msg->peer) will have been filled in by recvmmsg;
+             * for msg->local we parse the control data returned
+             */
+            if (BIO_MSG_N(msg, i).local != NULL)
+                if (extract_local(b, &mh[i].msg_hdr,
+                                  BIO_MSG_N(msg, i).local) < 1)
+                    return i > 0 ? (ossl_ssize_t)i : -1;
+        }
+    } else if (ret < 0) {
+        return PACK_ERRNO(get_last_socket_error());
+    }
+
+    return ret;
+
+#elif M_METHOD == M_METHOD_RECVMSG
+    /*
+     * If recvmsg is available, use it.
+     */
+    translate_msg(b, &mh, &iov, control, msg);
+
+    /* If local address was requested, it must have been enabled */
+    if (msg->local != NULL && !have_local_enabled)
+        /*
+         * If we have done at least one message, we must return the
+         * count; if we haven't done any, we can give an error code
+         */
+        return -1;
+
+    l = recvmsg(b->num, &mh, sysflags);
+    if (l < 0)
+        return PACK_ERRNO(get_last_socket_error());
+
+    msg->data_len   = (size_t)l;
+    msg->flags      = 0;
+
+    if (msg->local != NULL)
+        if (extract_local(b, &mh, msg->local) < 1) {
+            /*
+             * OS X exhibits odd behaviour where it appears that if a packet is
+             * sent before the receiving interface enables IP_PKTINFO, it will
+             * sometimes not have any control data returned even if the
+             * receiving interface enables IP_PKTINFO before calling recvmsg().
+             * This appears to occur non-deterministically. Presumably, OS X
+             * handles IP_PKTINFO at the time the packet is enqueued into a
+             * socket's receive queue, rather than at the time recvmsg() is
+             * called, unlike most other operating systems. Thus (if this
+             * hypothesis is correct) there is a race between where IP_PKTINFO
+             * is enabled by the process and when the kernel's network stack
+             * queues the incoming message.
+             *
+             * We cannot return the local address if we do not have it, but this
+             * is not a caller error either, so just return a zero address
+             * structure.
+             *
+             * We enable this workaround for Apple only as it should not
+             * be necessary otherwise.
+             */
+#if defined(__APPLE__)
+            memset(msg->local, 0, sizeof(*msg->local));
```

Comment:
Here we call memset, but later (line 1571 below) we call `BIO_ADDR_clear`. This looks odd.